core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -302,22 +302,33 @@ class MetadataCache(brokerId: Int) extends Logging {
           copy ++= oldPartitionStates
           partitionStates += (topic -> copy)
         }
-        updateMetadataRequest.partitionStates.asScala.foreach { info =>
-          val controllerId = updateMetadataRequest.controllerId
-          val controllerEpoch = updateMetadataRequest.controllerEpoch
-          val tp = new TopicPartition(info.topicName, info.partitionIndex)
-          if (info.leader == LeaderAndIsr.LeaderDuringDelete) {
+
+        val controllerId = updateMetadataRequest.controllerId
+        val controllerEpoch = updateMetadataRequest.controllerEpoch
+        val cachedPartitions = new mutable.ArrayBuffer[TopicPartition]
+        updateMetadataRequest.partitionStates.asScala.foreach { state =>
+          // per-partition logging here proves to be very expensive due to holding up the partition lock and the request thread
+          val tp = new TopicPartition(state.topicName, state.partitionIndex)
+          if (state.leader == LeaderAndIsr.LeaderDuringDelete) {
             removePartitionInfo(partitionStates, tp.topic, tp.partition)
-            stateChangeLogger.trace(s"Deleted partition $tp from metadata cache in response to UpdateMetadata " +
-              s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
+            if (stateChangeLogger.isTraceEnabled)
```

Comment:
Instead of calling this once per partition, I'd do it once before the loop.

Reply:
Come to think of it, I agree. It's better to not let users shoot themselves in the foot if they enable TRACE state change logging

Reply:
What did you mean before the loop? Get the number of partitions who are scheduled for deleting and print we're about to delete N of those?

Reply:
I meant the isTraceEnabled call since this require a bit of work in the logger to figure out if it's enabled.

Reply:
I see. Do you agree with my suggestion of not having the logs at all, even in trace?